File: movie-service/src/main/java/com/namnd/movieservice/service/MovieRatingService.java
package com.namnd.movieservice.service;

import com.namnd.movieservice.dto.CreateRatingRequest;
import com.namnd.movieservice.dto.MovieRatingDto;
import com.namnd.movieservice.dto.MovieRatingSummaryDto;

/**
 * Rating business logic: upsert and aggregation queries.
 */
public interface MovieRatingService {
    MovieRatingDto createOrUpdateRating(Long movieId, Long userId, CreateRatingRequest request);
    MovieRatingSummaryDto getRatingSummary(Long movieId, Long userId);
}

File: cinema-frontend/src/app/core/services/movie-rating.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { MovieRatingDto, MovieRatingSummaryDto, CreateRatingRequest } from '../models/movie-rating.model';

@Injectable({ providedIn: 'root' })
export class MovieRatingService {
  private http = inject(HttpClient);

  getRatingSummary(movieId: number): Observable<MovieRatingSummaryDto> {
    return this.http.get<MovieRatingSummaryDto>(`/api/movies/${movieId}/ratings`);
  }

  rateMovie(movieId: number, request: CreateRatingRequest): Observable<MovieRatingDto> {
    return this.http.post<MovieRatingDto>(`/api/movies/${movieId}/ratings`, request);
  }
}
